infra/relayer/src/purse.js

```diff
@@ -0,0 +1,696 @@
+    }
+  }
+
+  /**
+   * Get the gas price to use for a transaction
+   * @returns {BN} The gas price in gwei
+   */
+  async gasPrice() {
+    let gp = new BN('3000000000', 10) // 3gwei
+    try {
+      const netGp = stringToBN(await this.web3.eth.getGasPrice())
+      if (netGp) {
+        gp = netGp
+      }
+    } catch (err) {
+      logger.warn('Failed getting network gas price')
+      logger.debug(err)
+    }
+
+    if (gp.gt(MAX_GAS_PRICE)) {
+      // TODO: best way to handle this?
+      throw new Error('Current gas prices are too high!')
+    }
+
+    return gp
+  }
+
+  /**
+   * Select an available account
+   * @returns {string} address of the available account
+   */
+  async getAvailableAccount() {
+    let resolvedAccount = null
+
+    /**
+     * Select the account with the lowest pending that's under the max allowed
+     * pending transactions.  If none vailable, twiddle our thumbs until one
+     * becomes available...  The order is also important here.  According to the
+     * HD wallet "standards" the lowest index/path should be uesd first.  That
+     * way, any wallets using the mnemonic in the future will discover all the
+     * children and their funds.  It is supposed to stop at the first one
+     * without a tx history.
+     */
+    do {
+      let lowestPending = this.maxPendingPerAccount
+      for (const child of this.children) {
+        const childBal = stringToBN(await this.web3.eth.getBalance(child))
+        if (
+          this.accounts[child].pendingCount < lowestPending &&
+          childBal.gt(MIN_CHILD_BALANCE)
+        ) {
+          lowestPending = this.accounts[child].pendingCount
+          resolvedAccount = child
+          logger.debug(`selecting account ${child}`)
+        }
+      }
+
+      if (resolvedAccount) {
+        break
+      } else {
+        logger.debug('waiting for an account to become available...')
+      }
+    } while (await tick())
+
+    return resolvedAccount
+  }
+
+  /**
+   * Sign a transaction using a specific child account
+   * @param address {string} - The address of the account to sign the tx
+   * @param tx {object} - The transaction object
+   * @returns {object} The signed transaction object from web3.js
+   */
+  async signTx(address, tx) {
+    this._enforceChild(address)
+    await this.init()
+
+    const privKey = this.accounts[address].wallet.getPrivateKey()
+    return await this.web3.eth.accounts.signTransaction(
+      tx,
+      bufferToHex(privKey)
+    )
+  }
+
+  /**
+   * Send a transaction from an available sender account.
+   *
+   * NOTE: This does not wait for the transaction to be mined!
+   *
+   * @param tx {object} - The transaction object, sans `from` and `nonce`
+   * @param onReceipt {function} - An optional callback to call when a receipt is found
+   * @returns {string} The transaction hash of the sent transaction
+   */
+  async sendTx(tx, onReceipt) {
+    const address = await this.getAvailableAccount()
```

Comment:
If the relayer receives N requests in parallel, I don't know much about the node event scheduler, but isn't it possible that all N requests would then resolve to the same address ?

I guess my underlying question is do we need to have some sort of locking mechanism and load balancing of accounts being picked in getAvailableAccount ?

Reply:
This was kind of one of my concerns(as noted in the file TODO at the top).  It wasn't super clear to me if express would be handling parallel requests, which would then send those requests to Purse.  Seems likely, but I'm really not sure.  Since node doesn't support threads and all, unless there's multiple workers being stood up in the clusters(which would introduce even further issues with Purse), I'm thinking these requests should still be sequential in that there probably isn't a race condition here.  But my confidence in that isn't very high.  This might warrant further research.

Reply:
I've read a handful of articles and docs on the NodeJS event loop and I still can't seem to get a straight answer on this.  However, I think the solution is relatively simple and low-effort that I'm just going to go ahead and implement a simple lock that is only relevant between the `getAvailableAccount()` and `incrementTxCount()` calls.  

The worst that could happen is I spend 15 minutes to make Purse future-proof if we make any improvements to this Express server(e.g. [cluster](https://nodejs.org/api/cluster.html)).  At the best, we save ourselves some head-aches later down the road.

Will implement this later tonight or in the morning.  Thanks for raising this concern.

Reply:
Added a simple locking mechanism, and some gc in the process loop just in case...  Take a look when you get a chance.

Reply:
Based on my understanding the node event loop may switch to executing a new callback anytime there is some blocking IO. So in our case anytime there is a disk or network operation. So yes I'm pretty sure that if express receives concurrent requests, it would not process them sequentially but would switch from one to the other when an IO operation occur.
Makes sense to implement locking.